### ISSUE TITLE
PEP 681 + attrs + field specifiers

### ISSUE THREAD

**Issue by Tinche** [microsoft/pyright#5632]
Hello!

I'm trying to debug an issue with attrs typechecking.

Here's a dataclass example that works properly:

```python
from dataclasses import dataclass, field


@dataclass
class A:
    number: int = field(default="")  # Error, correct
```

and here's the attrs version, which doesn't report an error as far as I can tell:

```python
from attrs import define, field


@define
class A:
    number: int = field(default="")  # No error
```
(I'm using VS Code with pyright in strict mode.)

Is this an issue with pyright or are we doing something wrong in how we support PEP 681 in attrs? To save you a little searching, this is our code for the latest version that's on PyPI: https://github.com/python-attrs/attrs/blob/1e2f6f9cac5cc60f0adab051c14adf09ffe39155/src/attr/__init__.pyi#L115

**Comment by Tinche** [comment]
Alright, thanks. Why is that overload being used here though?

Just calling `field(default="")` ostensibly should produce a string, hitting a different overload.

![Screenshot 2023-08-02 at 19 14 39](https://github.com/microsoft/pyright/assets/1909233/aab1c324-9ab8-4d62-9bdf-2997cd89ac08)

**Comment by erictraut** [comment]
That overload doesn't satisfy the constraints of the call, so it produces an error. A type checker will continue to look for a subsequent overload that does not produce an error.

**Comment by Tinche** [comment]
Ok, I think I get it. Is there any up to date documentation on how pyright deals with choosing an overload? It seems to me it used to pick the first valid one by considering only the arguments, but that's no longer the case? Mypy handles this differently, but I'm sure you're already aware of this.

**Comment by erictraut** [comment]
Overload selection is extremely complicated. I've spend many hours trying to decipher mypy's approach, but I've found so many demonstrable inconsistencies in how it performs overload evaluation that I've given up trying to replicate it. It's also demonstrably deviates from its documented behavior.

I wish that mypy wouldn't have gone down the route of adding all of these baroque techniques for overload resolution. TypeScript's approach is much simpler (and preferable, IMO). It simply selects the first signature that matches the supplied arguments. It doesn't attempt to expand unions, nor does it change its behavior based on `Any` types in the argument list.

[Here](https://microsoft.github.io/pyright/#/type-concepts-advanced?id=overloads) is my attempt to document pyright's overload behavior, but even this documentation necessarily makes a bunch of assumptions about underlying implementation details.

Why is the last overload of `attrs.field` annotated to return `Any`? Should it not return `_T`?

**Comment by Tinche** [comment]
Thanks Eric, that was very helpful! I wasn't aware overload selection was that complex but between talking to you and some other folks I've learned something.

I didn't personally write these stubs, I'm just trying to fix them based on a Stack Overflow question. For the first step I wanted to understand exactly what's going on. Now that I have a better understanding I can work to fix the stubs.
